backend/dataall/modules/loader.py

```diff
@@ -29,49 +31,187 @@ class ModuleInterface(ABC):
 
 
-def _import_module(name):
+def _load_module(name: str):
+    """
+    Loads a module but not initializing it
+    """
     try:
         importlib.import_module(f"{_MODULE_PREFIX}.{name}")
         return True
     except ModuleNotFoundError:
         return False
+
+
+def _initialize_modules(modes: List[ImportMode]):
+    """
+    Initialize all modules for supported modes. This method is using topological sorting for a graph of module
+    dependencies. It's needed to load module in a specific order: first modules to load are without dependencies.
+    It might help to avoid possible issues if there is a load in the module constructor (which can be the case
+    if a module supports a few importing modes).
+    """
+    modules = _all_modules()
+    dependencies = defaultdict(list)
+    degrees = defaultdict(int)
+    supported = []
+    for module in modules:
+        if module.is_supported(modes):
+            supported.append(module)
+            degrees[module] += len(module.depends_on())
+            for dependency in module.depends_on():
+                dependencies[dependency].append(module)
+
+    queue = deque()
+    for module in supported:
+        if degrees[module] == 0:
+            queue.append(module)
+
+    initialized = 0
+    while queue:
+        to_init = queue.popleft()
+        _initialize_module(to_init)
+        initialized += 1
+
+        for dependant in dependencies[to_init]:
+            degrees[dependant] -= 1
+            if degrees[dependant] == 0:
+                queue.append(dependant)
+
+    if initialized < len(degrees):
+        # Happens if the ModuleInterface for dependency doesn't support import mode
+        # The case when there is circular imports should already be covered by python loader
+        raise ImportError("Not all modules have been initialized. Check if your import modes are correct")
+
+
+def _initialize_module(module):
+    module()  # call a constructor for initialization
+
+
+def _check_loading_correct(in_config: Set[str], modes: List[ImportMode]):
+    """
+    To avoid unintentional loading (without ModuleInterface) we can check all loaded modules.
+    Unintentional/incorrect loading might happen if module A has a direct reference to module B without declaring it
+    in ModuleInterface. Doing so, this might lead to a problem when a module interface require to load something during
+    initialization. But since ModuleInterface is not initializing properly (using depends_on)
+    some functionality may work wrongly.
+    """
+
+    expected_load = set(in_config)
+    for module in _all_modules():
+        for dependency in module.depends_on():
+            expected_load.add(dependency.name())
+
+    for module in _all_modules():
+        if module.is_supported(modes) and module.name() not in expected_load:
+            raise ImportError(
+                f"ModuleInterface has not been initialized for module {module.name()}. "
+                "Declare the module in depends_on"
```

Comment:
I am not sure that I follow all the logic here. In line 174 we are getting all the modules defined in_config (with active set to True or False) from `_load_modules`, shouldn't be only those that are active? I only see the `inactive` begin used in the `_describe_loading` function